### PR TITLE
cobalt/metrics: Wire detailed memory delegate and add unit tests

### PR DESCRIPTION
**Pull request by Awallky**
Wires CobaltDetailedMetricsDelegate into the memory instrumentation
service.

CobaltDetailedMetricsDelegate coordinates tracking of detailed process
metrics (such as Private Dirty, PSS, V8, Malloc, Stack, and executable
segments) for the browser and utility processes.

This change includes:
1. Implement CobaltDetailedMetricsDelegate to listen to client process
   memory dump notifications and classify metrics accordingly.
2. Add comprehensive unit tests in cobalt_detailed_metrics_delegate_unittest.cc
   verifying all memory classification, fallbacks, and edge cases.
3. Integrate CobaltDetailedMetricsDelegate in the Cobalt browser build.

Test: `run_cobalt_memory_instrumentation_unittests --gtest-filter="TracingObserverProtoTest.*:SmapsCategorizerTest.*:DetailedMetricsDelegateTest.*:SmapsMetricsTest.*:OSMetrics"`
Bug: 494004530

### CODE REVIEW

File: cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        }
+      }
+    }
+  }
+
+  base::AutoLock lock(lock_);
+  accumulated_stats_[std::string("pss:") + std::string(label)] +=
+      metrics.pss_kb;
+  accumulated_stats_[std::string("rss:") + std::string(label)] +=
+      metrics.rss_kb;
```
[the remark this answers is not in the document]
Comment (johnxwork) [reply cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc]: seems trivial to apply

[the remark this answers is not in the document]
Comment (Awallky) [reply cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc]: Done/

File: services/resource_coordinator/public/cpp/memory_instrumentation/os_metrics_linux.cc [hunk 2]
```diff
@@ -694,6 +1006,112 @@ std::vector<VmRegionPtr> OSMetrics::GetProcessMemoryMaps(
...
+  delegate->GetAndResetStats(&stats);
+  if (stats.size() <= 1) {
+    auto results = SmapsCategorizer::ScanSmaps(std::move(file_to_scan));
+    if (!results.has_value()) {
+      LOG(WARNING) << "Failed to read detailed metrics, continuing with basic metrics.";
+      return true;
+    }
+    for (const auto& entry : results.value()) {
+      delegate->OnSmapsEntry(entry.name, entry.metrics);
+    }
```
[the remark this answers is not in the document]
Comment (Awallky) [reply services/resource_coordinator/public/cpp/memory_instrumentation/os_metrics_linux.cc:1042]: Fixed.

File: cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace cobalt {
+
+namespace {
+
+// Ordered by probability/impact for early exit optimization.
+static constexpr CategoryPattern kCobaltPatterns[] = {
+    {"v8", "[anon:v8]", /*is_prefix=*/true},
+    {"partition_alloc", "[anon:partition_alloc]", /*is_prefix=*/true},
+    {"lib_chrobalt", "libchrobalt.so", /*is_prefix=*/false},
```
Comment (sherryzy) [inline cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc:29]: Do we actually have libchrobalt.so today? I thought it's always libcobalt.so for cobalt library?

File: cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    {"cobalt_core", "libcobalt.so", /*is_prefix=*/false},
```
Comment (sherryzy) [inline cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc]: Do we have both libcobalt.so and libchrobalt.so? What is cobalt_core?

**Review by sherryzy — COMMENTED** [review]
Overall looks good. Left two small questions,.

File: cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc (hunk 3 above)
Comment (Awallky) [reply cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc:29]: I think it varies. I have seen a mix of both in various build types. AFAICT all EG builds seem to use libcobalt.so.  However, ATV builds mostly use libchrobalt.so.

File: cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc (hunk 4 above)
Comment (Awallky) [reply cobalt/browser/metrics/cobalt_detailed_metrics_delegate.cc]: See other comment above about the use of the different names. But this is a good call out. I think we can just consolidate these into a single entity as they both represent our native library, depending on the build. 
I have changed so they map to the same name.

**Review by johnxwork — APPROVED** [review]
